src/win32/fs.cpp

```diff
@@ -1408,102 +1424,211 @@ void WinDirNotify::process(DWORD dwBytes)
+                if (!f) break;   // nullptr to cause the thread to exit
+                smQueue.pop_front();
+            }
+        }
+        if (f) 
+        {
+            f();
+            recheck = true;
+        }
+    }
+    LOG_debug << "Filesytem notify thread stopped";
+}
+
+WinDirNotify::WinDirNotify(string* localbasepath, string* ignore, WinFileSystemAccess* owner, Waiter* waiter) : DirNotify(localbasepath, ignore)
+{
+    fsaccess = owner;
+    fsaccess->dirnotifys.insert(this);
+    clientWaiter = waiter;
+
+    {
+        std::lock_guard<std::mutex> g(smNotifyMutex);
+        if (++smNotifierCount == 1)
+        {
+            smQueue.clear();
+            smEventHandle = CreateEvent(NULL, FALSE, FALSE, NULL);
+
+            // One thread to notify them all
+            smNotifierThread.reset(new std::thread([](){ notifierThreadFunction(); }));
+        }
+    }
+
 #ifndef WINDOWS_PHONE
     ZeroMemory(&overlapped, sizeof(overlapped));
     overlapped.hEvent = this;
-    enabled = false;
-    exit = false;
-    active = 0;
-
-    notifybuf[0].resize(65534);
-    notifybuf[1].resize(65534);
+    mOverlappedEnabled = false;
+    mOverlappedExit = false;
 
     int added = WinFileSystemAccess::sanitizedriveletter(localbasepath);
     localbasepath->append("", 1);
 
-    if ((hDirectory = CreateFileW((LPCWSTR)localbasepath->data(),
+    // ReadDirectoryChangesW: If you opened the file using the short name, you can receive change notifications for the short name.  (so make sure it's a long name)
+    std::wstring longname;
+    auto r = localbasepath->size() / sizeof(wchar_t) + 20;
+    longname.resize(r);
+    auto rr = GetLongPathNameW((LPCWSTR)localbasepath->data(), (LPWSTR)longname.data(), DWORD(r));
+
+    longname.resize(rr);
+    if (rr >= r)
+    {
+        rr = GetLongPathNameW((LPCWSTR)localbasepath->data(), (LPWSTR)longname.data(), rr);
+        longname.resize(rr);
+    }
+    localbasepath->resize(localbasepath->size() - added - 1);
+
+    if ((hDirectory = CreateFileW((LPCWSTR)longname.data(),
                                   FILE_LIST_DIRECTORY,
                                   FILE_SHARE_READ | FILE_SHARE_WRITE,
                                   NULL,
                                   OPEN_EXISTING,
                                   FILE_FLAG_BACKUP_SEMANTICS | FILE_FLAG_OVERLAPPED,
                                   NULL)) != INVALID_HANDLE_VALUE)
     {
-        failed = 0;
-        readchanges();
+        setFailed(0, "");
+
+        {
+            std::lock_guard<std::mutex> g(smNotifyMutex);
+            smQueue.push_back([this](){ readchanges(); });
+        }
+        SetEvent(smEventHandle);
     }
     else
     {
-        failed = GetLastError();
-        failreason = "CreateFileW was unable to open the folder";
-        LOG_err << "Unable to initialize filesystem notifications. Error: " << failed;
+        int err = GetLastError();
+        setFailed(err, "CreateFileW was unable to open the folder");
```

Comment:
It could be useful to add the `longname` to the reason here: _CreateFileW was unable to open the folder \<longname\>_ Actually, maybe better to be added to the log message below :)

Reply:
Agreed, but it's not super simple to do so currently- we need to convert the string from wchar to char first, and also the path will have been logged a few lines above with the sync creation logging.  When we introduce the LocalPath class, there will be a simple easy to use function to make that wchar to char conversion, let's add that logging then rather than spend time on it now and change it later.   thanks!